processing/src/main/java/io/druid/collections/spatial/Node.java

```diff
@@ -56,7 +57,7 @@ public Node(float[] minCoordinates, float[] maxCoordinates, boolean isLeaf, Bitm
   public Node(
       float[] minCoordinates,
       float[] maxCoordinates,
-      List<Node> children,
+      @Nullable Node child,
```

Comment:
There are two reasons why I chose not to pass a List.
 1. "It might be needed in the future" is BDUF, while Druid is mostly YAGNI in it's spirit, I believe. Also I doubt that this future will come, because the spatial subsystem didn't receive any changes at least in the last several years.
 2. The constructor was defined to accept a `List`, which was stored in a field of `Node` directly and possibly modified later. This violates the Effective Java's principle of making defensive copies on the API side, and was really confusing: when I was preparing this PR and saw the constructor of `Node` called with `Lists.newArrayList()` argument, I initially replaced that with `Collections.emptyList()`. This is wrong and was caught only in tests (fortunately!). But making defensive copies of lists seems to be wasteful for me in this case, as long as this is not really needed (see 1.)

Reply:
On YAGNI vs BDUF: IMO, YAGNI is better for internal APIs (which this is) so it is appropriate to get rid of the list. BDUF has some value for external APIs that we can't change later without disrupting users. It can pay to try to think about how usage may evolve over time, before the APIs are introduced.

So my 2¢ is that as an internal API, it's better for this one to offer no more than we need today.

Reply:
I don't have a strong opinion here, but I think anyone would wonder why this `Node` accepts only a single child in its constructor if he is not familiar with this part because they would know what the tree is and usually a node can have multiple children. I think that BDUF and YAGNI are the concepts to give some basic guidance and we should be able to apply it flexibly. 

Anyway, if you want to keep this, please add a comment why it currently accepts a single child.

Reply:
Added a comment